shared/utils/lazy.hpp

```diff
@@ -0,0 +1,74 @@
+#include <atomic>
+#include <cstdint>
+#include <functional>
+#include <optional>
+
+/// @brief Thread-safe container which ensures its value is initialized only once.
+/// @tparam T Type of the contained value.
+template <typename T>
+class Once {
+    std::optional<T> value;
+    std::atomic_uintptr_t state_and_queue;
+
+    void initialize(std::function<T()> const& initializer);
+
+  public:
+    /// @brief Constructs an uninitialized `Once<T>`.
+    constexpr Once() : value(), state_and_queue(0) {}
+
+    Once(Once<T> const&) = delete;
+
+    /// @brief Gets the current value, if it is initialized.
+    std::optional<T const&> get() const {
+        return value;
+    }
+    //// @brief Gets the current value, if it is initialized.
+    std::optional<T&> get() {
+        return value;
```

Comment:
I also imagine that because this is just returning the value of the optional, there is a potential for it to not yet be initialized, no? Though I don't know if that will just fallback `nullopt` or be some weird garbage